flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2392,12 +2392,14 @@ class ScalarArrayExprLowering {
     // 2) Lower the lhs expression to an array_update.
     semant = ConstituentSemantics::ProjectedCopyInCopyOut;
     auto lexv = lower(lhs);
-    // 3) Thread the array value updated forward
-    auto oldInnerArg =
-        mlir::cast<fir::ArrayUpdateOp>(fir::getBase(lexv).getDefiningOp())
-            .sequence();
-    auto offset = expSpace.argPosition(oldInnerArg);
-    expSpace.setInnerArg(offset, fir::getBase(lexv));
+    // 3) Thread the array value updated forward. Note: the lhs might be
```

Comment:
I'm having a hard time parsing this comment - "Thread the array value updated forward."

Reply:
All the array values at the current deepest level of the loop nest that may be updated are passed into the Region as block arguments. An array update is a transaction on that input value that produces a new value. Subsequent operations in the loop body need to use the updated value and not the original/stale one. This always has to happen because there will be a future `fir::result` operation added at minimum, which takes the value being threaded forward here and creates the phi node back to the start of the loop.